File: MongoBooks2/AngularMongoBooks3/ClientApp/src/app/Models/countryAuthors.ts
import { Author } from './author'

export interface ICountryAuthors
{
  name: string;
  totalBooksReadInLanguage: number;
  totalPagesReadInLanguage: number;
  totalBooksWorldWide: number;
  totalPagesWorldWide: number;
  percentageOfBooksRead: number;
  percentageOfPagesRead: number;
  flagUrl: string;
  authors: Author[];
};

export class CountryAuthors implements ICountryAuthors
{
  constructor(
    public name: string = "",
    public totalBooksReadInLanguage: number = 0,
    public totalPagesReadInLanguage: number = 0,
    public totalBooksWorldWide: number = 0,
    public totalPagesWorldWide: number = 0,
    public percentageOfBooksRead: number = 0,
    public percentageOfPagesRead: number = 0,
    public flagUrl: string = "",
    public authors: Author[] = new Array<Author>())
  {
  }

  public authorsNames: string;

  public static getAuthorsAsHtmlList(item: CountryAuthors): string
  {
    var htmlList: string = "<ul style=\"list-style-type:none;\">";

    if (item.authors != null && item.authors.length > 0) {
      for (let author of item.authors) {
        htmlList += "<li>" + author.name + "</li>";
      }
    }

    htmlList += "</ul>";

    return htmlList;
  }
}

File: MongoBooks2/AngularMongoBooks3/Controllers/DataClasses/CountryAuthors.cs
namespace AngularMongoBooks3.Controllers.DataClasses
{
    using System;
    using BooksCore.Books;

    public class CountryAuthors
    {
        public string Name { get; set; }

        public int TotalBooksReadInLanguage { get; set; }

        public int TotalPagesReadInLanguage { get; set; }

        public int TotalBooksWorldWide { get; set; }

        public int TotalPagesWorldWide { get; set; }

        public float PercentageOfBooksRead { get; set; }

        public float PercentageOfPagesRead { get; set; }

        public string FlagUrl { get; set; }

        public Author[] Authors { get; set; }

        public CountryAuthors()
        {

        }

        public CountryAuthors(AuthorCountry authorCountry)
        {
            Name = authorCountry.Country;
            TotalBooksReadInLanguage = authorCountry.TotalBooksWorldWide;
            TotalPagesReadInLanguage = (int)authorCountry.TotalPagesWorldWide;
            TotalBooksWorldWide = authorCountry.TotalBooksWorldWide;
            TotalPagesWorldWide = (int)authorCountry.TotalPagesWorldWide;
            PercentageOfBooksRead = (float)Math.Round(authorCountry.PercentageOfBooksRead, 2);
            PercentageOfPagesRead = (float)Math.Round(authorCountry.PercentageOfPagesRead, 2);
            FlagUrl = authorCountry.DisplayImage.ToString();

            Authors = new Author[authorCountry.AuthorsFromCountry.Count];
            for (int i = 0; i < authorCountry.AuthorsFromCountry.Count; i++)
            {
                Authors[i] = new Author(authorCountry.AuthorsFromCountry[i]);
            }
        }
    }
}
